### PR TITLE
libobs/UI: Add ability to copy/paste sources & filters

### PR DESCRIPTION
**Pull request by cg2121**

### CODE REVIEW

**Comment by Fenrirthviti** [comment]
Might be worth having two paste options, one as Paste as New and Paste as Existing to avoid issues with devices like capture cards being added multiple times.

**Comment by cg2121** [comment]
This is basically the same as adding an existing source in the add source dialog, so webcams are copied without problems.

**Comment by Fenrirthviti** [comment]
That's good for the default, but folks might also want to copy a source WITHOUT being used as an existing source. It has been requested several times the ability to duplicate a source without having to use add existing. Just a thought!

**Comment by WizardCM** [comment]
Also, how would this behave if someone tries copying from one Scene Collection to another?

**Comment by SuslikV** [comment]
@Fenrirthviti then we need 3 options:

- Copy
- Paste as Original
- Paste as Clone

Paste as Original always inserts original - if Paste as Original of the cloned source then insert original source.

---------------------
Or even for options:

- Copy as Clone
- Paste as Clone
- Copy
- Paste

And all combinations should work:

1. Copy as Clone --> Paste as Clone
2. Copy as Clone --> Paste (always inserts original)
3. Copy --> Paste as Clone
4. Copy --> Paste (always inserts original)

And this is too complicated for the users, I think.

**Comment by Heufneutje** [comment]
The 3 options seem the most user-friendly to me.

**Comment by WizardCM** [comment]
Why not

* Paste (as Existing)
* Paste (as Duplicate)

or something? "Clone" and "Original" could be reversible and mean the same thing.

**Comment by cg2121** [comment]
If someone tries to paste to another scene collection, nothing will happen.

**Comment by cg2121** [comment]
I have made it now that the paste is disabled when switching scene collections.

**Comment by cg2121** [comment]
I think having two paste options would confuse users. If users want the same source (clone), they should add the source the normal way.

**Comment by cg2121** [comment]
I've now added the ability to copy filters between sources.

File: libobs/obs-source.c [hunk 1]
```diff
@@ -391,6 +391,17 @@ static void duplicate_filters(obs_source_t *dst, obs_source_t *src,
 	da_free(filters);
 }
 
+void copy_paste_filters(obs_source_t *dst, obs_source_t *src)
+{
+	bool create_private;
+
+	duplicate_filters(dst, src, create_private ?
```
Comment (jp9000) [inline libobs/obs-source.c]: Careful here, variable was used while uninitialized.  This should have been caught by your compiler.

**Comment by jp9000** [comment]
Modifying this to allow pasting a reference or a full duplicate.

**Comment by jp9000** [comment]
Also, I ensured that if filters are pasted twice, that they don't have duplicate names.

**Pull request closed without merging**